### ISSUE TITLE
[Refactoring]: Generalize layers internally

### ISSUE THREAD

**Issue by kwvanderlinde** [RPTools/maptool#4378]
### Describe the problem

There's a lot of places in the codebase where we depend on specific layers to modify functionality. This makes it difficult to add new layers or think about other options for layering.

### The improvement you'd like to see

I would like to generalize layer functionality by defining a set of concrete properties that each layer provides. Functionality that today depends on specific layers would be rewritten to operate on any layers with matching properties. Also, functionality that today enumerates all layers by name would be rewritten to enumerate them another way (e.g., via `Layer.values()`).

I've identified the following properties for layers (all boolean):
1. Is it visible to players? _Currently `layer != GM`_
2. Is it a token layer? _Currently `layer == TOKEN`_
3. Is it a stamp layer _Currently `layer != TOKEN`_
4. Should STG tokens be snapped according to their center? _Currently `layer != BACKGROUND`, otherwise snapping by top-left corner._
5. Should token "drag"" using arrow keys be completed each step? _Currently `layer == BACKGROUND`_

Also a couple defaults that can be set in a common place rather than explicilty mentioned the layer everywhere:
1. `TOKEN` is a default/fallback layer for lots of player-oriented things.
2. `BACKGROUND` is the fallback layer for drawings.

### Expected Benefits

Layers will be more general, making it easier to define new layers without having to make code changes in a million places.

### Additional Context

The mentioned properties only _mostly_ define layer behaviour. There are a couple other bits that I can't see how to generalize right now, at least not in a satisfying way.

1. The `ZoneRenderer` has some very particular needs with layers, especially interleaving layers with each other (Hidden and Token) and other elements (Object and the grid).
2. When changing a token's layer ...
    - to Object or Background: non-figure tokens are changed to Top-down.
    - to Token: we try to guess a good shape, usually leaving Figures alone.
    - to GM: do nothing
    - BUT if we do it via macro functions instead of the UI, we don't treat Figures specially, and the "to GM" case is the same as "to Token".
3. The `getToken()` function defaults to using the Token and Hidden layer which otherwise have almost no properties in common. So how do we - and should we - generalize that?

**Comment by Azhrei** [comment]
I've thought about (1) previously.  It's obviously a policy problem, so delegate it to another function, probably a method of a yet-to-be-created `GameSystem` object.

The best I have come up with is to break down the components on each layer (tokens, token footprints, drawings, templates, auras, lights, vision outline, soft fog, hard fog, pathing) and create a `GameSystem.drawOrder()` method which returns an iterator of which elements to draw from which layer.  The iterator would delegate to a method that draws that component to a passed in graphics context and a "render context" object.  The latter is used to allow caching of intermediate results values between calls and/or for the next frame.

I don't have the ZoneRenderer code in front of me, so I likely missed something in the list of drawable components, and there may be other quirks as well.

That `GameSystem` object can handle some of these other duties as well.  For (2), a similar delegate of the `GameSystem` would handle layer-to-layer transitions.  Parameters could be the component being changed (token, drawable, template) and the new layer.  That method could then operate as required by the `GameSystem`.  For macro use, the method could have a third Boolean parameter, `noConstraints`, which moves the element to the new layer but doesn't otherwise tweak it.  Moving elements between layers could also be adjusted per the grid.

And last, for (3), delegate to the `GameSystem` object again.  Pass all existing parameters and let the `GameSystem` decide.

I'm not so sure about the latter two (haven't thought about them as much).  Hopefully, the above would allow for large scale battle games such as Axis and Allies, and a properly coded `GameSystem` could support teams and observers, rather than just GM & players.  (Although your idea for more generalized Token type than just PC and NPC would be helpful for teams.)

**Comment by FullBleed** [comment]
>     4. Should STG tokens be snapped according to their center? _Currently `layer != BACKGROUND`, otherwise snapping by top-left corner._

Would like to see this be less limited by layer and moved to a per token setting.

**Comment by Azhrei** [comment]
As related in #4379, it may be preferable for layers to have a flag allowing tokens on selected layers to be accessed by the owner(s) for movement and vision.  How about:
* bool ownerCanMove - allows owner to move token as well as cut/copy/paste

This allows a token to be moved to an "Invisible" layer and still be manipulated by the owner.  (Note that pathing would have to be drawn on the same layer so that only tokens that can see the contents of the invisible layer would be able to see the path.)

Not sure what to do about vision.  In the other FREQ, I had considered putting invisible tokens on a layer that wouldn't be seen by other tokens.  But tokens on that layer shouldn't necessarily be able to see each other either.  So now we're stuck with maintaining a list of which tokens can see the tokens on the invisible layer.  We also have to create multiple invisible layers if we want to simulate a multistory map layout in the future.

**Comment by kwvanderlinde** [comment]
> I've thought about (1) previously. It's obviously a policy problem, so delegate it to another function, probably a method of a yet-to-be-created GameSystem object.

That is an interesting thought. Just having a place for this logic to call home would help a lot. I'm still hoping that there is something intrinsic to the layers that can determine some of this behaviour, but we'll see if that works out. I'm willing to settle :smile: 

***

For the other comments, new features should be discussed in their own FRs. I do hope the internal representation of layers will be better able to support some of these new features, but beyond that goal this isn't the place for hashing out details on new features.

**Comment by thelsing** [comment]
I hope that we don't make MT too complicated for new users. We currently have already quite many settings and methods how to do things. If we move away from the clear layers each with its use case to a "configure your layers as you want" it might be a bit too much.

**Comment by Pmofmalasia** [comment]
> I hope that we don't make MT too complicated for new users.

I agree, but I think keeping the current layers as defaults and then having the option to add or remove layers would keep the complexity the same for new users. Maybe hide the layer stuff on an "Advanced" button underneath the layer selection or something similar.

**Comment by Azhrei** [comment]
> For the other comments, new features should be discussed in their own FRs.

Yep.  Only brought it up in case it has an impact here.

> I hope that we don't make MT too complicated for new users.

It's always been the plan to have some kind of "game system" preconfiguration for users.  They could choose D&D 5e or Star Wars or whatever, and the various settings would update to be appropriate for that game system.  (Craig added the dropdown list to Campaign Properties, for example.)  This means support will partially depend on "which game system are you using?" in order to understand their environment, but that's better than the scatter-gather Q&A we go through now about various settings.
